pkg/api/unversioned/types.go

```diff
@@ -54,6 +54,38 @@ type ListMeta struct {
 	ResourceVersion string `json:"resourceVersion,omitempty"`
 }
 
+// ListOptions is the query options to a standard REST list/watch calls.
+type ListOptions struct {
+	TypeMeta `json:",inline"`
+
+
+	// FIXME: It would be super convinient to have LabelSelector and FieldSelector
+	// of types: labels.Selector and fields.Selector respectively (no additional parsing
+	// in random places of the code).
+	//
+	// However, this would require (at least);
+	// - adding support for serialization/deserialization labels/fields.Selector to/from JSON
+	//   [my feeling is that MarshalJSON and UnmarshalJSON is what needs to be implemented for them]
+	// - adding some special treatment of those fields in convert to URL params - this would be
+  //   a dedicated, non-generic code (I think).
```

Comment:
We will probably change field selector conversion rules somewhere in the
future because we haven't actually defined how they work.  We may change
the names of query parameters to be consistent.  Those are my two examples
of possible changes to ListOptions.  We will probably remove "timeout".

On Nov 12, 2015, at 4:26 AM, Wojciech Tyczynski <notifications@github.com>
wrote:

In pkg/api/unversioned/types.go
<https://github.com/kubernetes/kubernetes/pull/17047#discussion_r44635903>:

> @@ -54,6 +54,38 @@ type ListMeta struct {
>  	ResourceVersion string `json:"resourceVersion,omitempty"`
>  }
>
> +// ListOptions is the query options to a standard REST list/watch calls.
> +type ListOptions struct {
> +	TypeMeta `json:",inline"`
> +
> +
> +	// FIXME: It would be super convinient to have LabelSelector and FieldSelector
> +	// of types: labels.Selector and fields.Selector respectively (no additional parsing
> +	// in random places of the code).
> +	//
> +	// However, this would require (at least);
> +	// - adding support for serialization/deserialization labels/fields.Selector to/from JSON
> +	//   [my feeling is that MarshalJSON and UnmarshalJSON is what needs to be implemented for them]
> +	// - adding some special treatment of those fields in convert to URL params - this would be
> +  //   a dedicated, non-generic code (I think).

Indentation on last line is wrong ;)

The idea was to make a decision and remove this comment :)

If you implement the text marshaller interfaces, will that work for both
url.Value and json marshalling? I think it's probably possible to make it
work without writing custom codec code.

This will not work for current implementation (i.e. it will not work for
url.Value).
However, after looking into this code, I think that it should be possible
to changing the code converting to/from url.Value to make it work.
@lavalamp <https://github.com/lavalamp> - so you think that having those
fields as "labels/fields.Selector" (which mean different that in v1 api) is
ok, right?

@smarterclayton <https://github.com/smarterclayton> may want to argue for
copying these to every group/version that uses them. I think we can do that
when we have versions that start needing different options.

I don't get it - are we still talking about ListOptions or something more
generic? If the former - I can't see usecase for different list options in
different apigroup. And if we need different options for different versions
- we shouldn't make then unversioned, right?
So to summarize - I'm a bit lost after this comment.

—
Reply to this email directly or view it on GitHub
<https://github.com/kubernetes/kubernetes/pull/17047/files#r44635903>.